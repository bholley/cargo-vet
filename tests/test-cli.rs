// These tests largely just check that basic CLI configs still work,
// and show you how you've changed the output. Yes, a lot of these
// will randomly churn (especially the help message ones, which
// contain the latest version), but this is good for two reasons:
//
// * You can easily see exactly what you changed
// * It can prompt you to update any other docs
//
// `cargo insta` automates reviewing and updating these snapshots.
// You can install `cargo insta` with:
//
// > cargo install cargo-insta
//
// Also note that `cargo test` for an application adds our binary to
// the env as `CARGO_BIN_EXE_<name>`.

use std::{
    path::PathBuf,
    process::{Command, Output, Stdio},
};

// Some tests need to write files (and read them back).
// To keep this tidy and hidden, we make a new directory
// in `target`.
// const TEST_TMP: &str = "../target/testdata/";

fn format_outputs(output: &Output) -> String {
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    // NOTE: We filter out the "Blocking: waiting for file lock" lines, as they
    // are printed out non-deterministically when there is file contention.
    let stderr = std::str::from_utf8(&output.stderr)
        .unwrap()
        .lines()
        .filter(|line| !line.starts_with("Blocking: waiting for file lock"))
        .collect::<Vec<_>>()
        .join("\n");
    format!("stdout:\n{stdout}\nstderr:\n{stderr}")
}

#[test]
fn test_version() {
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .arg("vet")
        .arg("-V")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    let stdout = String::from_utf8(output.stdout).unwrap();
    let stderr = String::from_utf8(output.stderr).unwrap();

    assert!(output.status.success(), "{}", stderr);
    assert_eq!(stderr, "");

    let (name, ver) = stdout.split_once(' ').unwrap();
    assert_eq!(name, "cargo-vet");
    let mut ver_parts = ver.trim().split('.');
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    ver_parts.next().unwrap().parse::<u8>().unwrap();
    assert!(ver_parts.next().is_none());
}

#[test]
fn test_long_help() {
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .arg("vet")
        .arg("--help")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("long-help", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_short_help() {
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .arg("vet")
        .arg("-h")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("short-help", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_markdown_help() {
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .arg("vet")
        .arg("help-markdown")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("markdown-help", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_json() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--output-format=json")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-json", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_suggest() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("suggest")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-suggest", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_suggest_json() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("suggest")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--output-format=json")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-suggest-json", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_suggest_shallow() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("suggest")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--shallow")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-suggest-shallow", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_suggest_shallow_json() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("suggest")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--output-format=json")
        .arg("--shallow")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-suggest-shallow-json", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_dump_graph_full_json() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("dump-graph")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--output-format=json")
        .arg("--depth=full")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-dump-graph-full-json", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_dump_graph_full() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("dump-graph")
        .arg("--diff-cache")
        .arg("../diff-cache.toml")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("--depth=full")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-dump-graph-full", format_outputs(&output));
    assert!(output.status.success(), "{}", output.status);
}

#[test]
fn test_project_bad_certify_human() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("certify")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("asdfsdfs")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-bad-certify-human", format_outputs(&output));
    assert!(!output.status.success(), "{}", output.status);
}

#[test]
fn test_project_bad_certify_json() {
    let project = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("test-project");
    let bin = env!("CARGO_BIN_EXE_cargo-vet");
    let output = Command::new(bin)
        .current_dir(&project)
        .arg("vet")
        .arg("certify")
        .arg("--output-format=json")
        .arg("--manifest-path")
        .arg("Cargo.toml")
        .arg("asdfsdfs")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();

    insta::assert_snapshot!("test-project-bad-certify-json", format_outputs(&output));
    assert!(!output.status.success(), "{}", output.status);
}
