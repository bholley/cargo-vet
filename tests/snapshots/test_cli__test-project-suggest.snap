---
source: tests/test-cli.rs
expression: format_outputs(&output)
---
stdout:
recommended audits for safe-to-deploy:
    cargo vet inspect tinyvec_macros 0.1.0                                             (used by tinyvec)                                           (81 lines)
    cargo vet inspect matches 0.1.9                                                    (used by url, idna, form_urlencoded)                        (210 lines)
    cargo vet inspect foreign-types-shared 0.1.1                                       (used by foreign-types)                                     (302 lines)
    cargo vet inspect try-lock 0.2.3                                                   (used by want)                                              (380 lines)
    cargo vet inspect openssl-probe 0.1.5                                              (used by native-tls)                                        (442 lines)
    cargo vet inspect tower-service 0.3.1                                              (used by hyper)                                             (501 lines)
    cargo vet inspect wasm-bindgen-shared 0.2.80                                       (used by wasm-bindgen-backend, wasm-bindgen-macro-support)  (515 lines)
    cargo vet inspect pin-utils 0.1.0                                                  (used by futures-util)                                      (538 lines)
    cargo vet inspect futures-sink 0.3.21                                              (used by h2, tokio-util)                                    (544 lines)
    cargo vet inspect cfg-if 1.0.0                                                     (used by log, instant, openssl, and 5 others)               (582 lines)
    cargo vet inspect foreign-types 0.3.2                                              (used by openssl)                                           (583 lines)
    cargo vet inspect remove_dir_all 0.5.3                                             (used by tempfile)                                          (592 lines)
    cargo vet inspect instant 0.1.12                                                   (used by fastrand)                                          (672 lines)
    cargo vet inspect hyper-tls 0.5.0                                                  (used by reqwest)                                           (673 lines)
    cargo vet inspect form_urlencoded 1.0.1                                            (used by url, serde_urlencoded)                             (689 lines)
    cargo vet inspect want 0.3.0                                                       (used by hyper)                                             (697 lines)
    cargo vet inspect percent-encoding 2.1.0                                           (used by url, reqwest, form_urlencoded)                     (702 lines)
    cargo vet inspect fnv 1.0.7                                                        (used by h2, http)                                          (730 lines)
    cargo vet inspect itoa 1.0.1                                                       (used by http, hyper, serde_json, serde_urlencoded)         (789 lines)
    cargo vet inspect lazy_static 1.4.0                                                (used by reqwest, schannel, and 3 others)                   (876 lines)
    cargo vet inspect httpdate 1.0.2                                                   (used by hyper)                                             (995 lines)
    cargo vet inspect tokio-native-tls 0.3.0                                           (used by reqwest, hyper-tls)                                (1089 lines)
    cargo vet inspect winapi-util 0.1.5                                                (used by termcolor)                                         (1095 lines)
    cargo vet inspect winapi-i686-pc-windows-gnu 0.4.0                                 (used by winapi)                                            (1132 lines)
    cargo vet inspect winapi-x86_64-pc-windows-gnu 0.4.0                               (used by winapi)                                            (1167 lines)
    cargo vet inspect futures-core 0.3.21                                              (used by h2, hyper, reqwest, and 3 others)                  (1175 lines)
    cargo vet inspect futures-task 0.3.21                                              (used by futures-util)                                      (1180 lines)
    cargo vet inspect http-body 0.4.4                                                  (used by hyper, reqwest)                                    (1256 lines)
    cargo vet inspect wasm-bindgen-macro 0.2.80                                        (used by wasm-bindgen)                                      (1274 lines)
    cargo vet inspect wasm-bindgen-futures 0.4.30                                      (used by reqwest)                                           (1303 lines)
    cargo vet inspect fastrand 1.7.0                                                   (used by tempfile)                                          (1368 lines)
    cargo vet inspect mime 0.3.16                                                      (used by reqwest)                                           (1715 lines)
    cargo vet inspect pkg-config 0.3.25                                                (used by openssl-sys)                                       (1747 lines)
    cargo vet inspect core-foundation-sys 0.8.3                                        (used by core-foundation, and 2 others)                     (1965 lines)
    cargo vet inspect wasm-bindgen-macro-support 0.2.80                                (used by wasm-bindgen-macro)                                (1965 lines)
    cargo vet inspect security-framework-sys 2.6.1                                     (used by native-tls, security-framework)                    (2016 lines)
    cargo vet inspect unicode-xid 0.2.2                                                (used by syn, proc-macro2, proc-macro2)                     (2044 lines)
    cargo vet inspect serde_urlencoded 0.7.1                                           (used by reqwest)                                           (2120 lines)
    cargo vet inspect termcolor 1.1.3                                                  (used by clap)                                              (2578 lines)
    cargo vet inspect slab 0.4.6                                                       (used by h2)                                                (2615 lines)
    cargo vet inspect os_str_bytes 6.0.0                                               (used by clap)                                              (2911 lines)
    cargo vet inspect wasm-bindgen-backend 0.2.80                                      (used by wasm-bindgen-macro-support)                        (3018 lines)
    cargo vet inspect miow 0.3.7                                                       (used by mio)                                               (3129 lines)
    cargo vet inspect unicode-bidi 0.3.7                                               (used by idna)                                              (3217 lines)
    cargo vet inspect wasi 0.11.0+wasi-snapshot-preview1                               (used by mio)                                               (3302 lines)
    cargo vet inspect native-tls 0.2.10                                                (used by reqwest, hyper-tls, tokio-native-tls)              (3715 lines)
    cargo vet inspect once_cell 1.10.0                                                 (used by openssl)                                           (3766 lines)
    cargo vet inspect quote 1.0.18                                                     (used by syn, and 4 others)                                 (3838 lines)
    cargo vet inspect redox_syscall 0.2.13                                             (used by tempfile)                                          (3890 lines)
    cargo vet inspect winreg 0.10.1                                                    (used by reqwest)                                           (3920 lines)
    cargo vet inspect core-foundation 0.9.3                                            (used by security-framework)                                (3953 lines)
    cargo vet inspect tracing-attributes 0.1.20                                        (used by tracing)                                           (3957 lines)
    cargo vet inspect ipnet 2.4.0                                                      (used by reqwest)                                           (3973 lines)
    cargo vet inspect futures-channel 0.3.21                                           (used by hyper)                                             (3992 lines)
    cargo vet inspect tempfile 3.3.0                                                   (used by native-tls)                                        (4059 lines)
    cargo vet inspect ryu 1.0.9                                                        (used by serde_json, serde_urlencoded)                      (4362 lines)
    cargo vet inspect schannel 0.1.19                                                  (used by native-tls)                                        (4601 lines)
    cargo vet inspect textwrap 0.15.0                                                  (used by clap)                                              (5196 lines)
    cargo vet inspect log 0.4.16                                                       (used by mio, want, reqwest, and 2 others)                  (5625 lines)
    cargo vet inspect proc-macro2 1.0.37@git:4445659b0f753a928059244c875a58bb12f791e9  (used by test-project)                                      (5691 lines)
    cargo vet inspect proc-macro2 1.0.37                                               (used by syn, quote, and 3 others)                          (5695 lines)
    cargo vet inspect socket2 0.4.4                                                    (used by hyper, tokio)                                      (6011 lines)
    cargo vet inspect pin-project-lite 0.2.8                                           (used by hyper, tokio, reqwest, and 4 others)               (6100 lines)
    cargo vet inspect tracing-core 0.1.25                                              (used by tracing)                                           (6156 lines)
    cargo vet inspect cc 1.0.73                                                        (used by openssl-sys)                                       (6554 lines)
    cargo vet inspect httparse 1.7.0                                                   (used by hyper)                                             (7258 lines)
    cargo vet inspect memchr 2.4.1                                                     (used by tokio, os_str_bytes)                               (8712 lines)
    cargo vet inspect bytes 1.1.0                                                      (used by h2, http, hyper, tokio, and 4 others)              (9207 lines)
    cargo vet inspect security-framework 2.6.1                                         (used by native-tls)                                        (9316 lines)
    cargo vet inspect indexmap 1.8.1                                                   (used by h2, clap)                                          (9445 lines)
    cargo vet inspect openssl-sys 0.9.72                                               (used by openssl, native-tls)                               (9477 lines)
    cargo vet inspect bumpalo 3.9.1                                                    (used by wasm-bindgen-backend)                              (10099 lines)
    cargo vet inspect tracing 0.1.33                                                   (used by h2, hyper, tokio-util)                             (10950 lines)
    cargo vet inspect mio 0.8.2                                                        (used by tokio)                                             (11719 lines)
    cargo vet inspect js-sys 0.3.57                                                    (used by reqwest, web-sys, wasm-bindgen-futures)            (12874 lines)
    cargo vet inspect tokio-util 0.7.1                                                 (used by h2)                                                (13870 lines)
    cargo vet inspect hashbrown 0.11.2                                                 (used by indexmap)                                          (14603 lines)
    cargo vet inspect serde 1.0.136                                                    (used by reqwest, serde_json, serde_urlencoded)             (16141 lines)
    cargo vet inspect url 2.2.2                                                        (used by reqwest)                                           (16344 lines)
    cargo vet inspect tinyvec 1.5.1                                                    (used by unicode-normalization)                             (16751 lines)
    cargo vet inspect http 0.2.6                                                       (used by h2, hyper, reqwest, http-body)                     (16819 lines)
    cargo vet inspect reqwest 0.11.10                                                  (used by test-project)                                      (19797 lines)
    cargo vet inspect wasm-bindgen 0.2.80                                              (used by js-sys, reqwest, web-sys, wasm-bindgen-futures)    (20977 lines)
    cargo vet inspect ntapi 0.3.7                                                      (used by mio)                                               (21227 lines)
    cargo vet inspect serde_json 1.0.79                                                (used by reqwest, test-project)                             (22848 lines)
    cargo vet inspect hyper 0.14.18                                                    (used by reqwest, hyper-tls)                                (24841 lines)
    cargo vet inspect futures-util 0.3.21                                              (used by h2, hyper, reqwest)                                (25067 lines)
    cargo vet inspect h2 0.3.13                                                        (used by hyper, reqwest)                                    (25417 lines)
    cargo vet inspect openssl 0.10.38                                                  (used by native-tls)                                        (28435 lines)
    cargo vet diff syn 1.0.0 1.0.91                                                    (used by tracing-attributes, and 2 others)                  (92 files changed, 21772 insertions(+), 6698 deletions(-))
    cargo vet inspect unicode-normalization 0.1.19                                     (used by idna)                                              (28622 lines)
    cargo vet inspect idna 0.2.3                                                       (used by url)                                               (32532 lines)
    cargo vet inspect vcpkg 0.2.15                                                     (used by openssl-sys)                                       (42642 lines)
    cargo vet inspect tokio 1.17.0                                                     (used by h2, hyper, reqwest, and 4 others)                  (91271 lines)
    cargo vet inspect libc 0.2.123                                                     (used by mio, atty, tokio, openssl, and 8 others)           (94060 lines)
    cargo vet inspect winapi 0.3.9                                                     (used by mio, atty, miow, ntapi, and 7 others)              (181323 lines)
    cargo vet inspect web-sys 0.3.57                                                   (used by reqwest, wasm-bindgen-futures)                     (197013 lines)
    cargo vet inspect encoding_rs 0.8.31                                               (used by reqwest)                                           (507245 lines)

recommended audits for safe-to-run:
    cargo vet inspect hermit-abi 0.1.19  (used by atty)  (932 lines)

estimated audit backlog: 1739832 lines

Use |cargo vet certify| to record the audits.

stderr:

